core/iwasm/common/wasm_c_api.c

```diff
@@ -449,6 +463,64 @@ wasm_engine_delete(wasm_engine_t *engine)
 #endif
 }
 
+#ifndef os_thread_local_attribute
+static unsigned
+retrive_thread_local_store_num(Vector *stores_by_tid, korp_tid tid)
+{
+    unsigned i = 0;
+
+    for (; i < stores_by_tid->num_elems; i++) {
+        thread_local_stores ts = { 0 };
+        bh_assert(bh_vector_get(stores_by_tid, i, &ts));
+
+        if (ts.tid == tid)
+            return ts.stores_num;
+    }
+
+    return 0;
+}
+
+static bool
+increase_thread_local_store_num(Vector *stores_by_tid, korp_tid tid)
+{
+    unsigned i = 0;
+    thread_local_stores ts = { 0 };
+
+    for (; i < stores_by_tid->num_elems; i++) {
+        bh_assert(bh_vector_get(stores_by_tid, i, &ts));
+
+        if (ts.tid == tid) {
+            ts.stores_num++;
+            bh_assert(bh_vector_set(stores_by_tid, i, &ts));
+            return true;
+        }
+    }
+
+    ts.tid = tid;
+    ts.stores_num = 1;
+    return bh_vector_append(stores_by_tid, &ts);
+}
+
+static void
+decrease_thread_local_store_num(Vector *stores_by_tid, korp_tid tid)
+{
+    unsigned i = 0;
+    thread_local_stores ts = { 0 };
+
+    for (; i < stores_by_tid->num_elems; i++) {
+        bh_assert(bh_vector_get(stores_by_tid, i, &ts));
+
+        if (ts.tid == tid) {
+            ts.stores_num--;
+            bh_assert(bh_vector_set(stores_by_tid, i, &ts));
```

Comment:
Should we remove ts from vector when stores_num is 0? Will there the case that stores_num is 0, and then we decrease it again?

Reply:
`engine->stores_by_tid` is a vector of `struct thread_local_stores` (instead of `therad_local_store*`), so I think it is ok to skip *remove*. Besides, I assume most of cases are about thread pool. If the counter is right, only meets *0* when quitting.
```
#ifndef os_thread_local_attribute
    if (!bh_vector_init(&engine->stores_by_tid, DEFAULT_VECTOR_INIT_SIZE,
                        sizeof(thread_local_stores), true))
        goto failed;
#endif
```

Reply:
OK, how about adding `bh_assert(ts.stores_num > 0);` before `ts.stores_num--;`?